content/blog/withr-3-0-0/index.Rmd

```diff
@@ -0,0 +1,228 @@
+## Cleaning up resources with base R and with withr
+
+Traditionally, resource cleanups in R is done with `base::on.exit()`. Cleaning up in the on-exit hook ensures that the cleanup happens both in the normal case, when the code has finished running without error, and in the error case, when something went wrong and execution is interrupted.
+
+`on.exit()` is meant to be used inside functions but it also works within `local()`, which we'll use here for our examples:
+
+```{r}
+local({
+  on.exit(message("Cleaning time!"))
+  1 + 2
+})
+```
+
+```{r, error = TRUE}
+local({
+  on.exit(message("Cleaning time!"))
+  stop("uh oh")
+  1 + 2
+})
+```
+
+`on.exit()` is guaranteed to run no matter what and this property makes it invaluable for resource cleaning. No more accidental littering!
+
+However the process of cleaning up this way can be a bit verbose and feel too manual. Here is how you'd create and clean up a temporary file for instance:
+
+```{r}
+local({
+  my_file <- tempfile()
+
+  file.create(my_file)
+  on.exit(file.remove(my_file))
+
+  writeLines(c("a", "b"), con = my_file)
+})
+```
+
+To streamline the process, withr provides `local_`-prefixed tools that combine both the creation or modification of a resource and its restoration to the original state in a single function:
+
+```{r}
+local({
+  my_file <- withr::local_tempfile()
+
+  writeLines(c("a", "b"), con = my_file)
+})
+```
+
+In this case we have created a resource (a file), but the same principle applies to modifying resources such as global options:
+
+```{r}
+local({
+  # Let's temporarily print with a single decimal place
+  withr::local_options(digits = 1)
+  print(1/3)
+})
+
+# The original option value has been restored
+getOption("digits")
+
+print(1/3)
+```
+
+And you can equivalently use the `with_`-prefixed variants (from which the package takes his name!), this way you don't need to wrap in `local()`:
+
+```{r}
+withr::with_options(list(digits = 1), print(1/3))
+```
+
+The `with_` functions are useful for creating very small scopes for given resources, inside or outside a function.
+
+
+## The withr 3.0.0 rewrite
+
+Traditionally, withr implemented its own exit event system on top of `on.exit()`. We needed an extra layer because of a couple of missing features:
+
+- When multiple resources are managed by a piece of code, the order in which these resources are restored or cleaned up sometimes matter. The most consistent order for cleanup is first in first out (FIFO). In other words the oldest resource, on which younger resources might depend, is cleaned up last.
+
+- The other missing piece was being able to inspect the contents of the exit hook. The `sys.on.exit()` R helper was created for this purpose but was affected by a bug that prevented it to work elsewhere than at top-level.
+
+We have contributed two changes to R 3.5.0 that filled these missing pieces. The `sys.on.exit()` bug was fixed and we've added an `after` argument to `on.exit()` to allow FIFO ordering.
```

Comment:
```suggestion
We contributed two changes to R 3.5.0 that filled these missing pieces, fixing the `sys.on.exit()` bug and adding an `after` argument to `on.exit()` to allow first-in first-out ordering.
```

Were they included in the 3.5 or 3.6 release?

Reply:
It was 3.5